src/tfs/tlog.c

```diff
@@ -202,31 +440,89 @@ closure_function(2, 1, void, log_read_complete,
     tl->dictionary = newdict;
 #endif
 
-    apply(sh, 0);
+  out_apply_status:
+    tlog_debug("log_read_complete exit with status %v\n", s);
+    buffer_clear(tl->tuple_staging);
+    apply(sh, s);
+  out:
     closure_finish();
 }
 
-void read_log(log tl, u64 offset, u64 size, status_handler sh)
+static boolean init_staging(log tl, status_handler sh)
 {
+    char *err;
+    u64 size = log_size(tl);
     tl->staging = allocate_buffer(tl->h, size);
+    if (tl->staging == INVALID_ADDRESS) {
+        err = "failed to allocate staging buffer";
+        goto fail;
+    }
+    tlog_debug("reading log extension, sectors %R, staging %p\n",
+               tl->sectors, tl->staging);
+
+#ifndef BOOT
+    /* reserve sectors in map */
+    if (!id_heap_set_area(tl->fs->storage, bytes_from_sectors(tl->fs, tl->sectors.start),
+                          size, true, true)) {
+        err = "failed to reserve sectors in allocation map";
+        goto fail;
+    }
+#endif
+    return true;
+  fail:
+    tlog_debug("%s\n", err);
+    apply(sh, timm("result", "%s", err));
+    deallocate_buffer(tl->staging);
+    tl->staging = 0;
+    return false;
+}
+
+void read_log(log tl, status_handler sh)
+{
+    assert(!tl->extension_open);
+    if (!init_staging(tl, sh))
+        return;
     status_handler tlc = closure(tl->h, log_read_complete, tl, sh);
-    range r = log_block_range(tl, tl->staging->length);
-//    rprintf("blocks %R\n", r);
-    apply(tl->fs->r, tl->staging->contents, r, tlc);
+    apply(tl->fs->r, tl->staging->contents, tl->sectors, tlc);
 }
 
 log log_create(heap h, filesystem fs, status_handler sh)
 {
     tlog_debug("log_create: heap %p, fs %p, sh %p\n", h, fs, sh);
     log tl = allocate(h, sizeof(struct log));
+    if (tl == INVALID_ADDRESS)
+        return tl;
     tl->h = h;
-    tl->offset = 0;
+    tl->sectors = irange(0, sector_from_offset(fs, TFS_LOG_DEFAULT_EXTENSION_SIZE));
     tl->fs = fs;
-    tl->completions = allocate_vector(h, 10);
+    tl->completions = allocate_vector(h, COMPLETION_QUEUE_SIZE);
+    if (tl->completions == INVALID_ADDRESS)
+        goto fail_dealloc_dict;
     tl->dictionary = allocate_table(h, identity_key, pointer_equal);
+    if (tl->dictionary == INVALID_ADDRESS)
+        goto fail_dealloc_tstage;
     tl->dirty = false;
     tl->staging = 0;
+    tl->tuple_staging = allocate_buffer(h, PAGESIZE /* arbitrary */);
```

Comment:
oh., I guess another reason for the write loop is that staging isn't the entire log, just a head-of-log cache?

oh..nevermind, I see it gets extended...I guess given the logic that's there it could be just the head now